### PR TITLE
libsyntax: remove dead code from parser.rs

### PR DESCRIPTION
**Pull request by matklad**
There is a dead code in libsyntax/parser/parse.rs, when parsing structs.

Two functions are involved:
- [parse_item_struct](https://github.com/rust-lang/rust/blob/cd9c9f048f6aa0be091cd9835771ba0712bead4e/src/libsyntax/parse/parser.rs#L4691)
- [parse_tuple_struct_body](https://github.com/rust-lang/rust/blob/cd9c9f048f6aa0be091cd9835771ba0712bead4e/src/libsyntax/parse/parser.rs#L4769)

The problem is that both functions handle the case with unit structs. But because
`parse_tuple_struct_body` is called from `parse_item_struct`, it never faces
this case.

This PR removes unit struct case from `parse_tuple_struct_body` function. I tested with `make -j8 check-statge1`.

### CODE REVIEW

**Comment by matklad** [comment]
Oups, the error message for erroneos code  `struct Foo for` changed, that's why the build is failing.

File: src/libsyntax/parse/parser.rs [hunk 1]
```diff
@@ -4726,9 +4726,13 @@ impl<'a> Parser<'a> {
...
             (fields, None)
         // Tuple-style struct definition with optional where-clause.
-        } else {
+        } else if self.token == token::OpenDelim(token::Paren) {
             let fields = try!(self.parse_tuple_struct_body(&class_name, &mut generics));
             (fields, Some(ast::DUMMY_NODE_ID))
+        } else {
+            let token_str = self.this_token_to_string();
+            return Err(self.fatal(&format!("expected `where`, `{}`, `(`, or `;` after struct \
+                                            name, found `{}`", "{", token_str)))
```
Comment (eddyb) [inline src/libsyntax/parse/parser.rs]: Shouldn't this be generated automatically, using the parser methods to check for the various tokens?
Also, `{` is escaped using `{{` in format strings.

Comment (matklad) [inline src/libsyntax/parse/parser.rs]: I don't really know for sure, I just moved this piece of code from `parse_tuple_struct_body`, here: https://github.com/matklad/rust/blob/3a360fca78b415d31783ec715a2c9024a89eeb75/src/libsyntax/parse/parser.rs#L4806-L4807 :)

If I just remove this explicit check and if test, then everything works, but the error message for one tests changes and the build fails.

Comment (matklad) [inline src/libsyntax/parse/parser.rs]: I'd say I can fix { escaping in this PR, in two places (there is a second similar error message in `parse_record_struct_body`).

 As for the removing the explicit error check, I'm not so sure:
* this PR does not change behavior at all, it just rearenges/removes code. Changes to code and test are better done in separate PR, if needed. 
* this eror message looks better then the autogenerated one.

Comment (eddyb) [inline src/libsyntax/parse/parser.rs]: Ah, I didn't realize it was pre-existing.

File: src/libsyntax/parse/parser.rs [hunk 2]
```diff
@@ -4769,43 +4773,32 @@ impl<'a> Parser<'a> {
...
-        // This is the case where we just see struct Foo<T> where T: Copy;
-        } else if self.token.is_keyword(keywords::Where) {
-            generics.where_clause = try!(self.parse_where_clause());
-            try!(self.expect(&token::Semi));
-            Ok(Vec::new())
-        // This case is where we see: `struct Foo<T>;`
-        } else {
-            let token_str = self.this_token_to_string();
-            Err(self.fatal(&format!("expected `where`, `{}`, `(`, or `;` after struct \
-                name, found `{}`", "{", token_str)))
```
Comment (eddyb) [inline src/libsyntax/parse/parser.rs]: This is where it previously was, I see.

File: src/libsyntax/parse/parser.rs (hunk 1 above)
Comment (matklad) [inline src/libsyntax/parse/parser.rs]: So should I fix the { issue?

Comment (eddyb) [inline src/libsyntax/parse/parser.rs]: Yes, it would be good to clear that up.

**Pull request merged**